Improve registration and verification flow

entantoencuanto opened the pull request:
#### :tophat: What? Why?

This PR improves the verification and registration flow when a user or visitant tries to make an action which requires at least one verification:
* Allows to store in the user extra data information about the action, resource with pending verifications. Some more data can be stored if required, the path to redirect after completing the verification flow and the permissions holder for actions not related with a specific resource of a component (for example, actions of initiatives or create resource action in a component)
* Changes the first login page to display the status of the verifications required to allow the pending action
* If a visitant is not signed in, after signing in or signing up and confirming email is redirected to the pending action onboarding page with the verifications required to authorize the action
* If there are more than one verification required to authorize an action after clicking on the button or link related to the action the user is redirected to the pending action onboarding and redirected to the same page after completing each verification if there are remaining pending verifications.
* After completing all the verifications if the action is authorized the user is redirected to the redirect path stored in its onboarding extended data or the resource path if the redirect path is missing with a success flash message
* If any check implies that the user does not have permission to perform the action, he is redirected to the resource with a forbidden action message.
* When the user browses the application and has an action pending of one or more authorizations a callout message appears with a link to complete the authorization flow
* The user is allowed to remove the data about the pending authorization flow if closes the callout message.
* Changes the authorized_to method to prepend a "Verify your account to ..." in the text of the button or link if not provided via block

#### :pushpin: Related Issues
*Link your PR to an issue*
- Fixes  #13038

#### Testing
* As admin edit the permissions of a resource (https://decidim-lot2.populate.tools/admin/participatory_processes/branch-highway/components/5/permissions/edit?resource_id=19&resource_name=meeting)
![image](https://github.com/user-attachments/assets/8c08e331-e653-4bea-8c8d-fe5f477be8ef)
* As user visit the resource (https://decidim-lot2.populate.tools/processes/branch-highway/f/5/meetings/19). If the action is not authorized yet the message should display a "Verify your account to..." When clicking on the action the user is redirected to the verifications flow page of the action (if there are more than one verification required to complete the action)
![image](https://github.com/user-attachments/assets/f58be7b2-0514-41d1-bded-eaad651ff383)
![Screenshot from 2024-09-05 18-07-30](https://github.com/user-attachments/assets/d97e121d-f274-48be-a49e-5c80d8905a06)

* If the user has an action with verifications pending a callout is shown with a link to the authorizations flow
![image](https://github.com/user-attachments/assets/55590344-b073-4852-9fcb-cece2d23b5e9)
* The authorizations page for the action shows the authorizations required to complete the action with the status of each one (https://decidim-lot2.populate.tools/authorizations/first_login). If the user gets all the pending authorizations or completes a verification and authorization is rejected the pending flow is removed and the user is redirected to the resource with a success / failure message
![Screenshot from 2024-09-05 18-53-12](https://github.com/user-attachments/assets/0eb50069-e80f-40d0-a68b-9ad547662940)
![Screenshot from 2024-09-05 18-54-20](https://github.com/user-attachments/assets/f2d38cd2-8207-4b12-9983-b4ecdf0b8931)

:hearts: Thank you!

ferblape commented on decidim-core/app/packs/src/decidim/onboarding/pending_action.js:
```diff
@@ -0,0 +1,19 @@
+        expires: 365
+      });
+
+      console.log("stored onboarding cookie", { action, model, permissionsHolder, redirectPath });
```
@entantoencuanto remove the console log

entantoencuanto replied:
Done!

ferblape commented on decidim-core/app/packs/src/decidim/onboarding/pending_action.js:
```diff
@@ -0,0 +1,19 @@
+      const redirectPath = element.dataset.onboardingRedirectPath;
+
+      Cookies.set("onboarding", JSON.stringify({ action, model, permissionsHolder, redirectPath }), {
+        expires: 365
```
Maybe move the 365 to a const in the top of the file?

ferblape commented on decidim-initiatives/app/helpers/decidim/initiatives/initiative_helper.rb:
```diff
@@ -30,44 +30,6 @@ def humanize_admin_state(state)
         I18n.t(state, scope: "decidim.initiatives.admin_states", default: :created)
       end
 
-      def authorized_create_modal_button(type, html_options, &)
```
👏🏾

ferblape commented on decidim-comments/config/locales/ca.yml:
```diff
@@ -38,7 +38,7 @@ ca:
     components:
       add_comment_form:
-        account_message: <a href="%{sign_in_url}">Inicia sessió</a> o <a href="%{sign_up_url}">crea un compte</a>per afegir el teu comentari.
+        account_message: Inicia sessió o crea un compte per afegir el teu comentari.
```
Why changing `ca` locale?

entantoencuanto replied:
There are some tests where the locale is changed to ca and es and this translation is in use. This produces an error because the old translations interpolate 2 values not provided by the changed view

ferblape commented on decidim-core/app/controllers/concerns/decidim/onboarding_action_methods.rb:
```diff
@@ -0,0 +1,43 @@
+      end
+
+      def store_onboarding_cookie_data!(user)
+        if cookies[:onboarding]
```
extract cookie name to constant?

entantoencuanto replied:
Done

ferblape commented on decidim-verifications/spec/system/authorizations_spec.rb:
```diff
@@ -297,6 +293,54 @@
+    let(:user) { create(:user, :confirmed, organization:, extended_data:) }
+    let(:commentable_path) { Decidim::ResourceLocatorPresenter.new(commentable).path }
+    let(:authorizations) { %w(dummy_authorization_handler dummy_authorization_workflow) }
+    # let!(:resource_permission) { commentable.create_resource_permission(permissions:) }
```
remove?

entantoencuanto replied:
Done and tests added

ferblape commented on decidim-core/app/services/decidim/onboarding_manager.rb:
```diff
@@ -0,0 +1,229 @@
+    # Removes the pending action from the user's extended_data.
+    #
+    # Returns nothing
+    def remove_pending_action!
```
This is duplicated in the concern `OnboardingActionMethods`. Given this service is used in a controller too, won't it be possible to avoid the duplication and use the concern method `clear_onboarding_data!(user)` to delete the pending action?

entantoencuanto replied:
You are right!

entantoencuanto commented on decidim-comments/app/cells/decidim/comments/comments/add_comment.erb:
```diff
@@ -3,13 +3,13 @@
-          sign_up_url: decidim.new_user_registration_path
-        ) %>
-      </span>
+      <div>
```
Not really, you are right

entantoencuanto commented on decidim-core/app/helpers/decidim/action_authorization_helper.rb:
```diff
@@ -48,6 +48,36 @@ def logged_button_to(*arguments, &block)
+        return action_text if authorization_status.blank?
+        return action_text if [:ok, :unauthorized].include?(authorization_status.global_code)
+
+        action_index = t("verify_to", scope: "decidim.core.actions", action: "%%%%").strip.index("%%%%")
```
Done!

entantoencuanto commented:
Hi, @NilHomedes , I think that it's ready to review in staging

NilHomedes reviewed (changes requested):
Hi @entantoencuanto, thanks for the job looks reaaly nice.

Here is my review with some change requests:

- [x]  We have been reviewing the different CTAs and copies affected by the request to change the action for "Verify your account to XXX". Finally, we think it's better to keep the original copy with the action (Vote, register, etc.). Therefore, we are removing the first AC of the issue. 
- [x]  As a participant, when I’m in the funnel, coming from: https://decidim-lot2.populate.tools/processes/branch-highway/f/4/proposals/1/
Once I have completed the authorizations needed, then I get an error by being redirected to: https://decidim-lot2.populate.tools/processes/branch-highway/f/4/proposals/1/proposal_vote?from_proposals_list=false
- [x]  Unexpected behaviour of the back button: As a registered participant, when I’m in the funnel, and I click on the back button I’m redirected to /Authorizations and I would expect to be redirected to the funnel page.
[AwesomeScreenshot-9_10_2024,3 46 35PM.webm](https://github.com/user-attachments/assets/3a8ee6a0-31dc-4a3a-9f57-1f7e069a9e41)
    
- [x]  Improve the copy of the funnel for: 
You're almost ready to vote on the "Test Proposal." Please verify your account by completing the following authorizations.

NilHomedes reviewed (changes requested):
Hi @entantoencuanto, I'm still getting the same errors I reported on my last review. See:

- [ ]  The CTA are still incorrect. We should delete the “Verify  your account to {action}” for the original copy. For example, in proposals the CTA should be “Vote” instead of “Verify your account to vote”. In comments, the text should be “Log in or create an account to add your comment.” Instead of “Verify your account to log in or create an account to add your comment”.
- [ ]  I’m still getting troubles when I complete all the authorizations coming from the funnel. I’m getting stuck here: https://decidim-lot2.populate.tools/processes/branch-highway/f/4/proposals/7/proposal_vote?from_proposals_list=false
It should work as follows: 
Given I’m redirected to the funnel coming from an action (vote on a proposal)
When I completed all the authorizations requested
Then I’m redirected to the proposal page, so I can complete the action (vote on that proposal).
- [ ]  Unexpected behaviour of the back button: As a registered participant, when I’m in the funnel, and I click on the back button I’m redirected to Authorizations (https://decidim-lot2.populate.tools/authorizations)  and I would expect to be redirected to the funnel page (https://decidim-lot2.populate.tools/authorizations/onboarding_pending).
- [ ]  The copy of the funnel has not been modified. As I said, we should improve the text for:
”You're almost ready to vote on the proposal "Test Proposal." Please verify your account by completing the following authorizations.”. 
I am referring to that copy:
![image](https://github.com/user-attachments/assets/5ccb82a5-31b3-4344-905b-9d0ba29c2e7f)

If you have any doubts, please do not hesitate to contact me. Thanks!

ferblape commented:
Sorry, the deploy failed, I'll ping you when it's correctly deployed

ferblape commented:
@NilHomedes the deployed finished successfully, could you please double check?

NilHomedes commented:
Hey @entantoencuanto 

I just found another aspect that could be improved. 
When only one verification is enabled, we are prompting the participant with that modal:
![Screenshot 2024-09-18 at 12-10-26 Do something - Test Site for Lot2](https://github.com/user-attachments/assets/93326101-84a9-407a-b7c4-df51f120dde5)

- [ ] We should skip that modal and send them directly to the verification form. 
To do that, we should improve some aspects of the form:
- [ ] Show an info announcement at the top of the verification form page. 
This announcement should say: 
"We need to verify your identity.  Please complete the form to proceed with this action."
![Authorization-methods](https://github.com/user-attachments/assets/876a7bc5-31c0-4ccd-8db7-42a698b743b3)

I also find another fix to do:
- [ ] Confirmation success message missing. After accepting the confirmation email, I should see a confirmation message saying that my email is confirmed. 

---
So, to sum-up. We will have 3 flows regarding the onboarding page:  
Flow 1: New participant
- Given I'm a new participant
When I accept the confirmation email
Then I'm redirected to the funnel (onboarding page)

Flow 2: Registered participant + 2 authorizations
- Given I'm a registered participant
When I click on a CTA with 2 or more authorizations enabled
Then, I'm redirected to the funnel (onboarding page)

Flow 3: Registered participant + 1 authorization
- Given I'm a registered participant
When I click on a CTA with ONE authorization enabled
Then, I'm redirected to verification form.


So, the funnel will only apply in flow 1 and 2.

entantoencuanto commented:
Sorry for the delay, @NilHomedes! The changes are deployed in the staging decidim-lot2 application. I have added all the requested changes.

The problem with the tests continues, although they all seem to pass, I have checked that there were errors. I have fixed all the tests in the files this PR touches but I don't know if there are any errors elsewhere.

entantoencuanto commented:
Hi, @NilHomedes, I have rebased the branch onto develop to fix the missing failing tests issue, but the result was the opposite, now there are a lot of failing tests related with ImageMagick that do not fail locally. The current rebased branch is deployed in staging. Despite the problems with tests, the changes can now be tested. Have you been able to review them?

NilHomedes reviewed (changes requested):
Hey @entantoencuanto!
Almost done, I just found a few things:

This is missing for the case of a registered participant.

> - [ ] Show an info announcement at the top of the verification form page.  This announcement should say: "We need to verify your identity. Please complete the form to proceed with this action."

And please, the announcement should be like in the image.
- [x] Remove the icon of the announcement
- [x] Use the flash warning color. 
- [x] Adjust the width of the announcement box to the form with

See the proto image:
![imatge](https://github.com/user-attachments/assets/c2318fdf-edd2-4dd3-90da-bcceb618c8b0)

entantoencuanto commented:
I have moved the message to the authorizations layout. Since each verification can define its own views and layouts in some cases the message may not appear and it should be included by the developers responsible of the flow. The previous removed message was included by a cell in the general layout of the application

NilHomedes commented:
Hi @entantoencuanto! 
I think we arte getting closer to the end of this PR.

Everything works fine from what I tested. There are only two things:

### Modify announcement copy
We notice that the announcement should be shorter, specially taking into account the mobile screen size. 
- [ ] So, to do that, we prefer to have an announcement that would work for all actions: 
"We need to verify your identity. 
Please complete the form to proceed with this action."

Instead of 
"We need to verify your identity.
Please complete the form to [action] in [resource-name]"

Usually, the `resource-name`could be so long and take too much space. 

### Error 
I'm getting an error when:
As a registered participant, I have one verificaton granted, and I try to join a [meeting](https://decidim-lot2.populate.tools/processes/branch-highway/f/5/meetings/19) that has 2 verifications enabled.
Then, I see the error: 

User ID: 59
Date and time: 2024-10-14T12:43:02.391645
Request method: GET
URL: https://decidim-lot2.populate.tools/authorizations/onboarding_pending
Reference: [c1633760-8d19-4c04-b09d-e29f2960dfbb]

entantoencuanto commented:
The error fix has not been deployed yet, I will let you know when changes are deployed.

entantoencuanto commented:
Hi @NilHomedes ! Tests reviewed and requested changes deployed. I think the reported bug is fixed, I tried to reproduce it and it didn't happen

NilHomedes reviewed (approved):
Yeeeah everything working good! Thanks for all the work @entantoencuanto ✨ 🙌 

Only a small thing I found in the authorizations forms, see:
![imatge](https://github.com/user-attachments/assets/4d6bdf99-2b3a-44a3-8b30-db3914195ae7)

Moving this to maintainers review.

entantoencuanto commented:
You are right! :smile: I have fixed in this PR, the typo was introduced in #13407

ferblape commented on decidim-core/app/services/decidim/onboarding_manager.rb:
```diff
@@ -0,0 +1,231 @@
+      onboarding_action
+    end
+
+    # Returns the translation of the action to be performed if found, action
+    # elsewhere
```
I think this comment is incomplete? (", action elsewhere" seems wrong)

entantoencuanto replied:
comment improved

ferblape commented on decidim-core/app/services/decidim/onboarding_manager.rb:
```diff
@@ -0,0 +1,231 @@
+    # Returns the resource title associated to the action. If the model is defined
+    # its title is used, if not the permissions holder title
+    #
+    # Returns a Hash
```
Hash or String?

entantoencuanto replied:
It's a translations Hash in general, but it can be a String in some cases. The returned value of this method is processed with `decidim_sanitize_translated`. I have expanded the comment

ferblape reviewed (dismissed):
Great job, just minor details on code and documentation

entantoencuanto commented on decidim-core/app/packs/src/decidim/onboarding_pending_action.js:
```diff
@@ -0,0 +1,22 @@
+import Cookies from "js-cookie";
+
+const ONBOARDING_COOKIE_EXPIRY = 365;
+const DATA_KEY = "onboarding";
```
Done

entantoencuanto commented on decidim-core/app/helpers/decidim/action_authorization_helper.rb:
```diff
@@ -57,52 +57,67 @@ def authorized_to(tag, action, arguments, block)
-        html_options["data-dialog-remote-url"] = modal_path(action, resource)
         url = "#"
+      elsif authorization_status&.ok? == false
+        html_options = clean_authorized_to_data_open(html_options.merge(onboarding_data_attributes(action, resource, permissions_holder, url, html_options.merge(tag:))))
```
I hope I have simplified the assignment somewhat. It is a very complex method.

entantoencuanto commented:
Hi, @alecslupu , I think this PR is ready to review. I'm working on a PR which depends on it and I'm going to start with other depending on the new one. Over the past few days I've added a few commits that fix bugs and solved conflicts with develop, but the overall behavior of the functionality has remained the same.

alecslupu reviewed (changes requested):
If the "example authorization" is being set like this: 
![image](https://github.com/user-attachments/assets/5b5bba02-6b86-49a3-a2ab-8afb50259c91)

Then the meeting page will raise the following error: 
![image](https://github.com/user-attachments/assets/31f2a987-27ff-4079-bc3b-d9830e614913)

entantoencuanto commented:
> If the "example authorization" is being set like this: ![image](https://private-user-images.githubusercontent.com/105683/388198082-5b5bba02-6b86-49a3-a2ab-8afb50259c91.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIxODc3MjcsIm5iZiI6MTczMjE4NzQyNywicGF0aCI6Ii8xMDU2ODMvMzg4MTk4MDgyLTViNWJiYTAyLTZiODYtNDlhMy1hMmFiLThhZmI1MDI1OWM5MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyMVQxMTEwMjdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00YzU3YzA3NzMwNTUwMjBjYjQyY2ExNGRlNmIyODhhNTUzZDVhZDc2MjYxNzk2NDg0NWZjOGEyNThiZTliNWQ0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.8qxuHmjOcn2GMoA8Oj28DVUaPiFuGWzXeu2SJWeL19Y)
> 
> Then the meeting page will raise the following error: ![image](https://private-user-images.githubusercontent.com/105683/388198248-31f2a987-27ff-4079-bc3b-d9830e614913.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIxODc3MjcsIm5iZiI6MTczMjE4NzQyNywicGF0aCI6Ii8xMDU2ODMvMzg4MTk4MjQ4LTMxZjJhOTg3LTI3ZmYtNDA3OS1iYzNiLWQ5ODMwZTYxNDkxMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyMVQxMTEwMjdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03Yjc3MjZhOGU1YWY1NmMxZTJhMDJmNzllMmVmN2YyZDYyMTJlYTM5Yjk2NzYzZDRlODAwNmY1Nzc1MGVmNDg1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.fqTB_yfjgKbbNZj104e6xegGF98yL-d_CR0ofJR5Yq0)

Yes, it's a bug of the dummy authorization handler when no scope is selected. I have fixed it in the next PR #13579 but I can cherry-pick the fix

entantoencuanto commented on decidim-comments/app/cells/decidim/comments/comment_cell.rb:
```diff
@@ -243,6 +243,10 @@ def vote_button_to(path, params, &)
         action_authorized_button_to(:vote_comment, path, params.merge(resource: root_commentable), &)
       end
+
+      def decidim_verifications
```
I thought the verifications module is required by the decidim core since the `Decidim::Authorization` model is associated with verifications. The functionality of this PR also depends on this verifications flow but I can investigate a way to disable it if the module is not installed.

alecslupu replied:
forget it for now

alecslupu commented:
While reviewing the functionality, when I have a meeting with external registratiions enabled, and when I click on "Register" i am being redirected to authorization screen, and after filling in the authorization data, i am being redirected to external website. 

IMO, if the meeting has an external registration url, maybe we could bypass the authorization? I would really like to know @decidim/product on this matter :)

entantoencuanto commented on decidim-verifications/app/views/decidim/verifications/id_documents/authorizations/choose.html.erb:
```diff
@@ -7,6 +7,6 @@
 <div class="mt-4 flex justify-between">
-  <%= link_to t("decidim.verifications.id_documents.authorizations.choose.offline"), { action: :new, using: :offline }, class: "button button__sm md:button__lg button__transparent-secondary" %>
-  <%= link_to t("decidim.verifications.id_documents.authorizations.choose.online"), { action: :new, using: :online }, class: "button button__sm md:button__lg button__secondary" %>
+  <%= link_to t("decidim.verifications.id_documents.authorizations.choose.offline"), { action: :new, using: :offline }.merge(redirect_url:).compact, class: "button button__sm md:button__lg button__transparent-secondary" %>
+  <%= link_to t("decidim.verifications.id_documents.authorizations.choose.online"), { action: :new, using: :online }.merge(redirect_url:).compact, class: "button button__sm md:button__lg button__secondary" %>
```
Totally agree :sweat_smile:

entantoencuanto commented on decidim-verifications/spec/system/action_authorization_spec.rb:
```diff
@@ -8,7 +8,7 @@
-    create(:organization, available_authorizations: [authorization])
+    create(:organization, available_authorizations:)
   end
 
```
Yes, it's clearer

alecslupu reviewed (approved):
LGTM! 

Merging with Codecov pipeline failed.
